src/permutation.nr

```diff
@@ -2,527 +2,238 @@ use crate::encoding::{
 
-    let choose_output = decode_choose(e.e, f.e, g.e);
+        let s1 = decode_xor(w[i - 2].w2_rotate_sum);
 
-    xor_output + choose_output
-}
-
-struct RoundState {
-    a: EncodedMajority,
-    b: EncodedMajority,
-    c: EncodedMajority,
-    d: EncodedMajority,
-    e: EncodedChoose,
-    f: EncodedChoose,
-    g: EncodedChoose,
-    h: EncodedChoose,
-}
-
-// TODO: initialize this via a constant method that adds no constraints?
-impl RoundState {
-    fn new() -> Self {
-        RoundState {
-            a: encode_a(SHA512_INITIAL_HASH_VALUES[0] as Field),
-            b: encode_a(SHA512_INITIAL_HASH_VALUES[0] as Field),
-            c: encode_a(SHA512_INITIAL_HASH_VALUES[1] as Field),
-            d: encode_a(SHA512_INITIAL_HASH_VALUES[2] as Field),
-            e: encode_e(SHA512_INITIAL_HASH_VALUES[3] as Field),
-            f: encode_e(SHA512_INITIAL_HASH_VALUES[4] as Field),
-            g: encode_e(SHA512_INITIAL_HASH_VALUES[5] as Field),
-            h: encode_e(SHA512_INITIAL_HASH_VALUES[6] as Field),
-        }
+        result[i] = reduce_modulo_64(w[i - 16].raw + s0 + w[i - 7].raw + s1);
     }
+    result
 }
 
-unconstrained fn __reduce_modulo_64(input: Field) -> (Field, Field) {
-    let input128 = input as u128;
-
-    let remainder = input128 & 0xffffffffffffffff;
-    let modulus = input128 >> 64;
-
-    (modulus as Field, remainder as Field)
-}
-
-fn reduce_modulo_64(input: Field) -> Field {
-    // Safety: obtain result and then validate result is correct
-    let (modulus, remainder) = unsafe { __reduce_modulo_64(input) };
-
-    modulus.assert_max_bit_size::<8>();
-    remainder.assert_max_bit_size::<64>();
-    assert_eq(modulus * 0x10000000000000000 + remainder, input);
-
-    remainder
-}
-
-fn sha512_compression(_input: INT_BLOCK, _state: STATE) -> STATE {
-    let mut message_input: [Field; 16] = [0; 16];
-    for i in 0..16 {
-        message_input[i] = _input[i] as Field;
-    }
-
-    let mut state: RoundState = RoundState {
-        a: encode_a(_state[0] as Field),
-        b: encode_a(_state[1] as Field),
-        c: encode_a(_state[2] as Field),
-        d: encode_a(_state[3] as Field),
-        e: encode_e(_state[4] as Field),
-        f: encode_e(_state[5] as Field),
-        g: encode_e(_state[6] as Field),
-        h: encode_e(_state[7] as Field),
-    };
-
-    let mut w: [Field; 80] = extend_witness(message_input);
-
-    compress(&mut state, w);
-
-    let mut output: STATE = [0; 8];
-    output[0] = state.a.raw as u64;
-    output[1] = state.b.raw as u64;
-    output[2] = state.c.raw as u64;
-    output[3] = state.d.raw as u64;
-    output[4] = state.e.raw as u64;
-    output[5] = state.f.raw as u64;
-    output[6] = state.g.raw as u64;
-    output[7] = state.h.raw as u64;
-    output
-}
-
-fn compress(state: &mut RoundState, w: [Field; 80]) {
+/**
+ * @brief evaluate the SHA384/512 compression function
+ * @details each iteration costs 243 gates
+ **/
```

Comment:
Please don't use `@brief` or `@details` as these aren't meaningful in Noir. Currently this is rendered as

> * @\brief evaluate the SHA384/512 compression function
> * @\details each iteration costs 243 gates

(minus the backslashes which are just to avoid tagging people)

 Idiomatic Noir follows the Rust documentation conventions so we use `///` triple slashes and a blank line between paragraphs otherwise things get pushed onto the same line.

```noir
/// Evaluate the SHA384/512 compression function
///
/// # Cost
/// Each iteration costs 243 gates
pub(crate) fn compress(state: &mut RoundState, w: [Field; 80]) {
```

Reply:
thanks for the clarification, i've been using doxygen markup as i was unaware of the standard we were using.

do you have a link to the rust documentation conventions you could share?

Reply:
Rust documentation standards are fairly loose. There's no special syntax beyond linking to different types within the crate and references for links (Noir doesn't support these currently so they can be ignored) so it's basically just markdown.


https://doc.rust-lang.org/rustdoc/how-to-write-documentation.html
https://doc.rust-lang.org/book/ch14-02-publishing-to-crates-io.html#commonly-used-sections

Reply:
I've updated all the doc comments to use the rust syntax

Reply:
Awesome, thank you 🙏